### PR TITLE
Fix setting `kubeContext` in skaffold

### PR DESCRIPTION
**Pull request by gsquared94**
Fixes: #5334, #4347 

**Description**

Skaffold has multiple ways to set the current cluster context and it does not work as intended. The main issues are:
- The global config tries to read the kubeconfig before the flags `--kube-context` and `--kubeconfig` have been processed, and stores a singleton result that doesn't get recomputed ever.
- The `runcontext` tries to figure out if the cluster is local or remote heuristically from the kubecontext but even this happens before the flags `--kube-context` and `--kubeconfig` have been processed
- These flags are only processed after all skaffold configs have been parsed and this is to support reading a third source of `kubecontext` from the skaffold config's `Deploy.KubeContext` field.

This PR fixes these issues and will result in the following state:
- Current set kubecontext is always overridden by the flags `--kube-context` and `--kubeconfig` and as soon as the command starts.
- We no longer treat the value provided in a skaffold config's `Deploy.KubeContext` as an override for the entire process' kubecontext value. Rather it only overrides the kubecontext provided to each deployer only at the time of deployment.
- This allows deploying to different kubecontexts in different skaffold modules.

Follow up work:
- Defining multiple `kubeContext` in different skaffold configs in a multi-config project is still disallowed with an artificial check as before, since it needs further changes to the `Logger` and `PortForwarder` constructs.

### CODE REVIEW

**Review by tejal29 — COMMENTED** [review]
> * Current set kubecontext is always overridden by the flags `--kube-context` and `--kubeconfig` and as soon as the command starts.
> * We no longer treat the value provided in a skaffold config's `Deploy.KubeContext` as an override for the entire process' kubecontext value. Rather it only overrides the kubecontext provided to each deployer only at the time of deployment.

If this is the case, should we even evaluate `kubecontext`at the start of the run via congifure.Once?
Could we not just create the runner with kubeContext either from CLI or current config selected by user. Once this is done for all command requiring deploys, skaffold code should only rely on `Runcontext.KubeContext` field?


> * This allows deploying to different kubecontexts in different skaffold modules (another requested feature)
> * This necessitates that the status checker is moved inside the deployer and has a single instance per deployer (another request #5774)  to handle the different kubecontexts.

File: cmd/skaffold/app/cmd/cmd.go [hunk 1]
```diff
@@ -95,6 +95,9 @@ func NewSkaffoldCommand(out, errOut io.Writer) *cobra.Command {
 				return err
 			}
 
+			// Setup kubeContext and kubeConfig
+			kubectx.ConfigureKubeConfig(opts.KubeConfig, opts.KubeContext)
+
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply cmd/skaffold/app/cmd/cmd.go:100]: unfortunately we evaluate the global config for all skaffold commands when we evaluate `config.ShouldDisplaySurveyPrompt`. We need to interpret the resolved `kubeContext` prior to evaluating the global config. So we need to do this check here also.

File: pkg/skaffold/kubernetes/context/context.go [hunk 2]
```diff
@@ -44,19 +44,15 @@ var (
 // When given, the firstCliValue always takes precedence over the yamlValue.
 // Changing the kube-context of a running Skaffold process is not supported, so
 // after the first call, the kube-context will be locked.
-func ConfigureKubeConfig(cliKubeConfig, cliKubeContext, yamlKubeContext string) {
-	newKubeContext := yamlKubeContext
-	if cliKubeContext != "" {
-		newKubeContext = cliKubeContext
-	}
+func ConfigureKubeConfig(cliKubeConfig, cliKubeContext string) {
 	configureOnce.Do(func() {
```
Comment (briandealwis) [inline pkg/skaffold/kubernetes/context/context.go:48]: In an ideal world, `pkg/skaffold` should be treated like a library, and these values should be computed in `cmd/skaffold` and provided into the library.

There are also the `KUBECONFIG` environment variable that should influencing this decision (and note that it is a `PATH`-like variable that can specify multiple files).

File: pkg/skaffold/kubernetes/context/context.go [hunk 3]
```diff
@@ -44,19 +44,15 @@ var (
...
+		kubeContext = cliKubeContext
 		kubeConfigFile = cliKubeConfig
 		if kubeContext != "" {
 			logrus.Infof("Activated kube-context %q", kubeContext)
 		}
 	})
-	if kubeContext != newKubeContext {
+	if kubeContext != cliKubeContext {
 		logrus.Warn("Changing the kube-context is not supported after startup. Please restart Skaffold to take effect.")
 	}
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply pkg/skaffold/kubernetes/context/context.go]: done

**Review by briandealwis — APPROVED** [review]

**Pull request merged**